sbi/simulators/simutils.py

```diff
@@ -2,82 +2,480 @@
+    """Raise ValueError if dimensionality of data doesn't match requirements.
     
-    Returns function name if no name is given. 
+    The current requirements are that is 2D with a leading batch dimension of size 1.
+    That is, neither multiple, batched observations are supported, 
+    nor a multidimensional observation e.g., an image.
+    """
+
+    if true_observation.squeeze().ndim > 1:
+        raise ValueError(
+            """`true_observation` has D>1 dimensions. SBI interprets the leading 
+                dimension as a batch dimension, but it *currently* only processes 
+                a single observation, i.e. the first element of the batch.
+                
+                For example:
+                
+                > true_observation = [ [1,2,3], [4,5,6] ] 
+                
+                is interpreted as two vector observations, only the first of which 
+                is currently used to condition inference.
+                
+                Use rather:
+                
+                > true_observation = [ [[1,2,3], [4,5,6]] ]
+                > true_observation = [ [1], [2], [3]]
+                
+                if your single observation is matrix-shaped or scalar-shaped.
+                
+                Finally:
+                
+                > true_observation = [1]
+                > true_observation = [1, 2, 3]
+                
+                will be interpreted as one scalar observation or one vector observation
+                and don't require wrapping (unsqueezing).
+                """
+        )
+
+
+def check_prior_methods(prior):
+    """Check whether the prior has methods .sample and .log_prob. 
     
-    Arguments:
-        simulator_fun {function} -- simulator function taking parameter batch as only argument, return data. 
+    Raises: 
+        AttributionError: if either of the two methods doesnt works as expected. 
+    """
+
+    # Sample a batch of two parameters to check batch behaviour > 1.
+    num_samples = 2
+    # Check .sample and log_prob, keep for later use.
+    try:
+        theta = prior.sample((num_samples,))
+    except AttributeError:
+        raise AttributeError(
+            "Prior needs method .sample((num_samples, )). Consider using a "
+            "pytorch Distribution"
+        )
+    try:
+        log_probs = prior.log_prob(theta)
+    except AttributeError:
+        raise AttributeError(
+            "Prior needs method .log_prob(values). Consider using a pytorch Distribution"
+        )
+
+
+def check_prior_batch_behavior(prior):
+    """Assert that it is possible to sample and evaluate batches of parameters."""
+
+    # Check for correct batch size in .sample and .log_prob
+    num_samples = 2
+    theta = prior.sample((num_samples,))
+    log_probs = prior.log_prob(theta)
+
+    assert (
+        len(theta.shape) == 2
+    ), f"""A parameter batch sampled from the prior must be 2D, 
```

Comment:
hm we havent discussed parameter matrices yet, do we want to handle them? Is it a realistic use-case to have a prior that returns matrices?

Reply:
There can be simulators which need matrices as parameters. Think anything depending e.g. on a covariance. 

Whatever we do it is important to not introduce undocumented limitations because we thought that the case would not appear in practice. 

Would there be a problem caused somewhere by demanding >=2? At the end, whether matrix thetas are supported is entirely the user's problem (prior and simulator have to support it). If it is too difficult, keep it `==2` but make clear that this is a limitation of our code. OTOH, if you don't limit this, we can add a test with matrix theta and see where it breaks. This would allow us to document accordingly, so that we can lift this in the future.

Reply:
No, no problem. Will use that and we will get the error at some point, create an issue and so on.